syne_tune/optimizer/schedulers/searchers/kde_searcher.py

```diff
@@ -0,0 +1,275 @@
+                return res
+            else:
+                return [(value - domain.lower) / (domain.upper - domain.lower)]
+
+        return np.hstack([
+            numerize(value=config[k], domain=v, categorical_map=categorical_maps.get(k, {}))
+            for k, v in configspace.items()
+            if isinstance(v, sp.Domain)
+        ])
+
+    @staticmethod
+    def from_feature(configspace, feature_vector, inv_categorical_maps):
+        def inv_numerize(values, domain, categorical_map):
+            if not isinstance(domain, sp.Domain):
+                # constant value
+                return domain
+            else:
+                if isinstance(domain, sp.Categorical):
+                    index = int(values * len(domain))
+                    return categorical_map[index]
+                else:
+                    if isinstance(domain, sp.Float):
+                        return values * (domain.upper - domain.lower) + domain.lower
+                    else:
+                        return int(values * (domain.upper - domain.lower) + domain.lower)
+
+        res = {}
+        curr_pos = 0
+        for k, domain in configspace.items():
+            if hasattr(domain, "sample"):
+                res[k] = domain.cast(
+                    inv_numerize(
+                        values=feature_vector[curr_pos],
+                        domain=domain,
+                        categorical_map=inv_categorical_maps.get(k, {})
+                    )
+                )
+                curr_pos += 1
+            else:
+                res[k] = domain
+        return res
+
+    def to_objective(self, result: Dict):
+        if self.mode == 'min':
+            return result[self.metric_name]
+        elif self.mode == 'max':
+            return -result[self.metric_name]
+
+    def _update(self, trial_id: str, config: Dict, result: Dict):
+        self.X.append(self.to_feature(
+            config=config,
+            configspace=self.configspace,
+            categorical_maps=self.categorical_maps,
+        ))
+        self.y.append(self.to_objective(result))
+        self.train_kde(np.array(self.X), np.array(self.y))
+
+    def get_config(self, **kwargs):
+        # if not done with points_to_evaluate, pick the next one
+        if self.points_to_evaluate is not None and self.num_evaluations < len(self.points_to_evaluate):
+            suggestion = self.points_to_evaluate[self.num_evaluations]
+        else:
+            if self.good_kde is None and self.bad_kde is None:
+                # if not enough suggestion made, sample randomly
+                suggestion = {k: v.sample()
+                if isinstance(v, sp.Domain) else v for k, v in self.configspace.items()}
+            elif np.random.rand() < self.random_fraction:
+                suggestion = {k: v.sample()
+                if isinstance(v, sp.Domain) else v for k, v in self.configspace.items()}
+            else:
+                l = self.good_kde.pdf
```

Comment:
One of these two could be None.

Reply:
Not sure what you mean? What could be none, the model?

Reply:
there are either both none or not

Reply:
OK